core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1817,8 +1834,182 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+    *
+    * @param servicePerimeterName
+    * @return
+    */
+  private def loadStaticProjectsForPerimeter(servicePerimeterName: ServicePerimeterName): Seq[GoogleProjectNumber] = {
+    config.staticProjectsInPerimeters.getOrElse(servicePerimeterName, Seq.empty)
+  }
+
+  /**
+    * Takes the the name of a Service Perimeter as the only parameter.  Since multiple Billing Projects can specify the
+    * same Service Perimeter, we will:
+    * 1. Load all the Billing Projects that also use this servicePerimeterName
+    * 2. Load all the Workspaces in all of those Billing Projects
+    * 3. Collect all of the GoogleProjectNumbers from those Workspaces
+    * 4. Post that list to Google to overwrite the Service Perimeter's list of included Google Projects
+    * 5. Poll until Google Operation to update the Service Perimeter gets to some terminal state
+    * Throw exceptions if any of this goes awry
+    *
+    * @param servicePerimeterName
+    * @return Future[Unit] indicating whether we succeeded to update the Service Perimeter
+    */
+  private def overwriteGoogleProjectsInPerimeter(servicePerimeterName: ServicePerimeterName): Future[Unit] = {
+    collectWorkspacesInPerimeter(servicePerimeterName).map { workspacesInPerimeter =>
+      val projectNumbers = workspacesInPerimeter.flatMap(_.googleProjectNumber) ++ loadStaticProjectsForPerimeter(servicePerimeterName)
+      val projectNumberStrings = projectNumbers.map(_.value)
+
+      // Make the call to Google to overwrite the project.  Poll and wait for the Google Operation to complete
+      gcsDAO.accessContextManagerDAO.overwriteProjectsInServicePerimeter(servicePerimeterName, projectNumberStrings).map { operation =>
+        // Keep retrying the pollOperation until the OperationStatus that gets returned is some terminal status
+        retryUntilSuccessOrTimeout(failureLogMessage = s"Google Operation to update Service Perimeter: ${servicePerimeterName} was not successful")(5 seconds, 50 seconds) { () =>
+          gcsDAO.pollOperation(OperationId(GoogleApiTypes.AccessContextManagerApi, operation.getName)).map {
+            case OperationStatus(false, _) => Future.failed(new RawlsException(s"Google Operation to update Service Perimeter ${servicePerimeterName} is still in progress..."))
+            // TODO: If the operation to update the Service Perimeter failed, we need to consider the possibility that
+            // the list of Projects in the Perimeter may have been wiped or somehow modified in an undesirable way.  If
+            // this happened, it would be possible for Projects intended to be in the Perimeter are NOT in that
+            // Perimeter anymore, which is a problem.
+            case OperationStatus(true, errorMessage) if !errorMessage.isEmpty => Future.successful(throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, s"Google Operation to update Service Perimeter ${servicePerimeterName} failed with message: ${errorMessage}")))
```

Comment:
Another thing to think about is the Projects per Perimeter limit.  What happens when we try to overwrite a Service Perimeter with 2001 projects if the limit is 2000 in the perimeter?  Will the whole request fail?  Will only the last record not be added to the perimeter?  Will exactly one project at random not be added to the perimeter?  What do we do about the project(s) that failed to get added?

Reply:
let it ride. I suspect the entire request will fail leaving the perimeter unchanged. The current published quota is 10k but Greg J. says it is actually 20k.